beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ESTIMATION_CLUSTER_BASED_ESTIMATOR_HPP
+#define BELUGA_ESTIMATION_CLUSTER_BASED_ESTIMATOR_HPP
+
+// standard library
+#include <algorithm>
+#include <initializer_list>
+#include <numeric>
+#include <optional>
+#include <queue>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+// external
+#include <range/v3/action/sort.hpp>
+#include <range/v3/numeric/accumulate.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/zip.hpp>
+#include <sophus/se2.hpp>
+#include <sophus/types.hpp>
+
+// project
+#include <beluga/algorithm/spatial_hash.hpp>
+#include <beluga/estimation/utils/range_statistics.hpp>
+#include <beluga/views.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a cluster-base estimator mixin.
+ */
+
+namespace beluga {
+
+namespace cse_detail {
+
+/// @brief A struct that holds the data of a single cell in the grid.
+struct GridCellData {
+  double weight{0.0};                         ///< average weight of the cell
+  std::size_t num_particles{0};               ///< number of particles in the cell
+  Sophus::SE2d representative_pose_in_world;  ///< state of a particle that is within the cell
+  std::optional<std::size_t> cluster_id;      ///< cluster id of the cell
+};
+
+/// @brief A map that holds the sparse data about the particles grouped in cells. Used by the clusterization algorithm.
+using GridCellDataMap2D = std::unordered_map<std::size_t, GridCellData>;
+
+/// @brief Function that creates an vector containing the hashes of each of the states in the input range.
+/// @tparam Range Type of the states range.
+/// @tparam Hasher Hash function type to convert states into hashes.
+/// @param states The range of states.
+/// @param spatial_hash_function_ The hash object instance.
+/// @return A vector containing the hashes of each of the states in the input range.
+template <class Range, class Hasher>
+[[nodiscard]] auto precalculate_particle_hashes(Range&& states, const Hasher& spatial_hash_function_) {
+  const auto state_to_range = [&](const auto& state) { return spatial_hash_function_(state); };
+  return states | ranges::views::transform(state_to_range) | ranges::to<std::vector<std::size_t>>();
```

Comment:
```suggestion
  return states | ranges::views::transform(spatial_hash_function_) | ranges::to<std::vector<std::size_t>>();
```

meta: If this is only used once, it might be better to write it in place, it isn't super hard to understand and there will be less documentation to mantain.

Reply:
I wrote in place, and declared for the three uses in the unit test.